### PR TITLE
Update the `disjoint sequences` when new block bodies arrive

### PR DESCRIPTION
**Pull request by rafal-ch**
Maintain the sequence of fully available blocks.

### CODE REVIEW

File: node/src/components/storage.rs [hunk 1]
```diff
@@ -1154,6 +1175,31 @@ impl Storage {
         })
     }
 
+    /// Returns `true` if there is a block body for the given block header available in storage.
+    fn has_corresponsing_body<Tx: Transaction>(
```
Comment (Fraser999) [inline node/src/components/storage.rs]:
```suggestion
    fn has_corresponding_body<Tx: Transaction>(
```

File: node/src/components/storage.rs [hunk 2]
```diff
@@ -1154,6 +1175,31 @@ impl Storage {
...
+    /// adds heights of blocks that share the same body.
+    fn update_disjoint_height_sequences(&mut self, block_header: &BlockHeader) {
+        self.disjoint_block_height_sequences
+            .insert(block_header.height());
+
+        if let Some(block_bodies_to_add) = self.missing_block_bodies.get(block_header.body_hash()) {
+            self.disjoint_block_height_sequences
+                .extend(block_bodies_to_add);
+            self.missing_block_bodies.remove(block_header.body_hash());
+        }
```
Comment (Fraser999) [inline node/src/components/storage.rs]:
```suggestion
        if let Some(block_bodies_to_add) = self.missing_block_bodies.remove(block_header.body_hash()) {
            self.disjoint_block_height_sequences
                .extend(block_bodies_to_add);
        }
```

File: node/src/components/storage/disjoint_sequences.rs [hunk 3]
```diff
@@ -19,8 +20,8 @@ enum InsertOutcome {
 }
 
 /// Represents a continuous sequence of `u64`s.
-#[derive(Copy, Clone, Debug, Eq, PartialEq, DataSize)]
-struct Sequence {
+#[derive(Copy, Clone, Debug, Eq, PartialEq, DataSize, Ord, PartialOrd)]
+pub(crate) struct Sequence {
```
Comment (Fraser999) [inline node/src/components/storage/disjoint_sequences.rs]: Can we restrict to `pub(super)`?

File: node/src/components/storage/tests.rs [hunk 4]
```diff
@@ -28,6 +28,7 @@ use crate::{
     },
     crypto::AsymmetricKeyExt,
     effect::{requests::StorageRequest, Multiple},
+    storage::disjoint_sequences::Sequence,
```
Comment (Fraser999) [inline node/src/components/storage/tests.rs]: We have all the other `storage` module imports above here in the `use super::{...}` block.

File: node/src/components/storage/tests.rs [hunk 5]
```diff
@@ -38,6 +39,23 @@ use crate::{
...
 type BlockGenerators = Vec<fn(&mut TestRng) -> (Block, EraId)>;
 
+impl Storage {
+    fn disjoint_sequences(&self) -> &Vec<Sequence> {
+        self.disjoint_block_height_sequences.sequences()
+    }
+
+    fn add_missing_block_body(&mut self, block_header: &BlockHeader) {
+        match self.missing_block_bodies.entry(*block_header.body_hash()) {
+            std::collections::hash_map::Entry::Occupied(mut entry) => {
```
Comment (Fraser999) [inline node/src/components/storage/tests.rs]: We should import at least `std::collections::hash_map`, or preferably `std::collections::hash_map::Entry`.

File: node/src/components/storage.rs (hunk 2 above)
Comment (rafal-ch) [reply node/src/components/storage.rs]: TIL :+1:

**Review by Fraser999 — APPROVED** [review]

**Pull request merged**